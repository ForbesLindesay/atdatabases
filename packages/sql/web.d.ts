// @autogenerated

import def from './lib/web';

export default def;
export * from './lib/web';
