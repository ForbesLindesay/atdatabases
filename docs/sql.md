---
id: sql
title: Building SQL Queries
---

All SQL Databases in @databases use the same approach for building SQL Queries. Using tagged template literals gives you a powerful and flexible way of creating queries without opening yourself to SQL Injection attacks.

We do not do the escaping of SQL ourselves, instead we pass the query and parameters separately to the underlying database engine. This is both more secure and more performant.

## API

```ts
import sql from '@databases/sql';
```

### ``` sql`...` ```

Builds part of, or the whoel of, an SQL query, safely interpretting the embedded expressions. You can use it for a basic SQL query with no parameters:

```ts
db.query(sql`SELECT * FROM users;`);
// => {text: 'SELECT * FROM users;', values: []}
```

You can interpolate values:

```ts
const id = 10;
db.query(sql`SELECT * FROM users WHERE id=${id};`);
// => {text: 'SELECT * FROM users WHERE id=$1;', values: [10]}
```

You can also interpolate other queries:

```ts
const FIELD_SET = sql`id, name`;
const id = 10;
db.query(sql`SELECT ${FIELD_SET} FROM users WHERE id=${id};`);
// => {text: 'SELECT id, name FROM users WHERE id=$1;', values: [10]}
```

### ``` sql.ident(...names) ```

If you want to dynamically include an identifier in a query, you can use `sql.identifier`. e.g.

```ts
const fieldName = 'id';
db.query(sql`SELECT ${sql.ident(fieldName)} FROM users;`);
// => {text: 'SELECT "id" FROM users;', values: []}
```

You can also specify identifiers with a schema/namespace using this approach.

```ts
const fieldName = sql.ident('u', 'id');
db.query(sql`SELECT ${fieldName} FROM users AS u;`);
// => {text: 'SELECT "u"."id" FROM users AS u;', values: []}
```

### ``` sql.value(val) ```

`sql.value(val)` acts as a shorthand for ``` sql`${val}` ```. It takes a value, and represents it with a placeholder in the resulting query.

```ts
const id = sql.value(10);
db.query(sql`SELECT * FROM users WHERE id=${id};`);
// => {text: 'SELECT * FROM users WHERE id=$1;', values: [10]}
```

### ``` sql.join(arrayOfFragments, delimiter) ```

Joins an array of SQLQuery values using the delimtier (which is treated as a raw SQL string). It properly handles joining the array of values and ensuring that the placeholders match up.

```ts
const arrayOfSqlFields = ["a", "b", "c", "d"].map(
  n => sql.identifier(n),
);

sql`SELECT ${sql.join(arrayOfSqlFields, ", ")}`;
// => {text: 'SELECT "a", "b", "c", "d";', values: []}

const arrayOfSqlConditions = [
  sql.query`a = ${1}`,
  sql.query`b = ${2}`,
  sql.query`c = ${3}`
];
sql`WHERE (${sql.join(arrayOfSqlConditions, ') AND (')})`;
// => {text: 'WHERE (a = $1) AND (b = $2) AND (c = $3)', values: [1, 2, 3]}
```

> N.B. the delimiter should always be a string literal, never allow user input to be passed as the delimiter. If you allow user specified delimiters, it will lead to an SQL Injection Vulnerability.

### ``` sql.__dangerous__rawValue(str) ```

This is an escape hatch to allow you to take a string from a source you trust and treat it as an SQL query. There is almost always a better way.

```ts
const idSet = sql.__dangerous__rawValue('(1, 2, 3)');
db.query(sql`SELECT * FROM users WHERE id in ${idSet};`);
// => {text: 'SELECT * FROM users WHERE id in (1, 2, 3);', values: []}
```

> N.B. if you pass user input to this function, it will result in an SQL Injection Vulnerability. That is why it has such an odd name.

### ``` sql.registerFormatter(class, format) ```

The `registerFormatter` function allows you to add custom handling for a given class. Instead of just directly adding the value to the "values" array, this lets you write your own function that takes the value and returns an `SQLQuery`.

```ts
function padNumber(value, length) {
  let str = `${value}`;
  while (str.length < length) {
    str = '0' + str;
  }
  return str;
}

class DayDate {
  constructor(year: number, month: number, day: number) {
    this.year = year;
    this.month = month;
    this.day = day;
  }
}

sql.registerFormatter(DayDate, (date) => sql.value(
  `${padNumber(date.year, 4)}-${padNumber(date.month, 2)}-${padNumber(date.day, 2)}`,
));


db.query(sql`SELECT * FROM users WHERE dob = ${new DayDate(2018, 1, 20)};`);
// => {text: 'SELECT * FROM users WHERE dob = $1;', values: ['2018-01-20']}
```

### ``` SQLQuery.compile() ```

This returns an object with `{text: string, values: any[]}` where the `text` field contains the SQL query formatted for postgres. It also minifies the query using `pg-minify` by default. You can pass `SQLQuery.compile({minify: false})` to disable minifying.

### ``` SQLQuery.compileMySql() ```

This returns an object with `{text: string, values: any[]}` where the `text` field contains the SQL query formatted for MySQL. We also use this function to generate queries for SQLite as it has a very similar format.
