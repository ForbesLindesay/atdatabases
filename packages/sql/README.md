# @databases/sql

For documentation, see https://www.atdatabases.org/docs/sql