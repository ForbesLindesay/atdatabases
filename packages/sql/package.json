{
  "name": "@databases/sql",
  "version": "0.0.0",
  "description": "",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "dependencies": {
    "pg-minify": "^0.5.5"
  },
  "devDependencies": {
    "@types/node": "^10.12.17"
  },
  "scripts": {},
  "repository": "https://github.com/ForbesLindesay/atdatabases/tree/master/packages/sql",
  "license": "GPL-3.0",
  "publishConfig": {
    "access": "public"
  },
  "bugs": "https://github.com/ForbesLindesay/atdatabases/issues"
}
