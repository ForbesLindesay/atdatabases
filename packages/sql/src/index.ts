import SQLQuery from './SQLQuery';

export {SQLQuery};

/**
 * The interface we actually expect people to use.
 */
export interface SQL {
  (strings: TemplateStringsArray, ...values: Array<any>): SQLQuery;

  join(queries: Array<SQLQuery>, seperator?: string): SQLQuery;
  __dangerous__rawValue(text: string): SQLQuery;
  value(value: any): SQLQuery;
  ident(...names: Array<any>): SQLQuery;
  registerFormatter<T>(
    constructor: new (...args: any[]) => T,
    format: (value: T) => SQLQuery,
  ): void;
}

// Create the SQL interface we export.
const modifiedSQL: SQL = Object.assign(
  (strings: TemplateStringsArray, ...values: Array<any>): SQLQuery =>
    SQLQuery.query(strings, ...values),
  {
    // tslint:disable:no-unbound-method
    join: SQLQuery.join,
    __dangerous__rawValue: SQLQuery.raw,
    value: SQLQuery.value,
    ident: SQLQuery.ident,
    registerFormatter: SQLQuery.registerFormatter,
    // tslint:enable:no-unbound-method
  },
);

export default modifiedSQL;

module.exports = modifiedSQL;
module.exports.default = modifiedSQL;
module.exports.SQLQuery = SQLQuery;
