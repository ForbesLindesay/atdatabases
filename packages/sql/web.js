// @autogenerated

module.exports = require('./lib/web');