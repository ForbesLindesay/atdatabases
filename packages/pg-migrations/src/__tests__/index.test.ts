import {readFileSync} from 'fs';
import connect from '@databases/pg';
import sql from '@databases/sql';
import {MigrationsPackage} from '../';
import buildPackage from '../build-package';

jest.setTimeout(30000);

// for some reason prettier in jest fails if this isn't required before it is used
require('prettier/parser-typescript');

let output: MigrationsPackage = null as any;

test('generate', async () => {
  await buildPackage({
    migrationsDirectory: __dirname + '/migrations',
    outputFile: __dirname + '/output.ts',
    databasesDbPgMigrationsName: '../',
  });
  expect(readFileSync(__dirname + '/output.ts', 'utf8')).toMatchSnapshot();
  output = require('./output').default;
  expect(output).toBeInstanceOf(MigrationsPackage);
});

const db = connect();

test('clean database', async () => {
  await db.query(sql`
    DROP TABLE IF EXISTS users;
    DROP TABLE IF EXISTS "atdatabases_migrations";
    DROP TABLE IF EXISTS "atdatabases_migrationsVersion";
  `);
});
test('upAll', async () => {
  await output.upAll(undefined, {silent: true});
  expect(await db.query(sql`SELECT * FROM users`)).toMatchSnapshot();
});
test('downLast', async () => {
  await output.downLast(undefined, {silent: true});
  expect(await db.query(sql`SELECT * FROM users`)).toMatchSnapshot();
  await output.downLast(undefined, {silent: true});
  expect(await db.query(sql`SELECT * FROM users`)).toMatchSnapshot();
  await output.upAll(undefined, {silent: true});
});
async function getError(fn: () => any) {
  try {
    await fn();
  } catch (ex) {
    return ex;
  }
}
test('downOne', async () => {
  await output.downOne(undefined, {silent: true});
  expect(await db.query(sql`SELECT * FROM users`)).toMatchSnapshot();
  await output.downOne(undefined, {silent: true});
  const err = await getError(
    async () => await db.query(sql`SELECT * FROM users`),
  );
  expect(() => {
    if (err) throw err;
  }).toThrowErrorMatchingSnapshot();
  await output.upAll(undefined, {silent: true});
});
test('downAll', async () => {
  await output.downAll(undefined, {silent: true});
  const err = await getError(
    async () => await db.query(sql`SELECT * FROM users`),
  );
  expect(() => {
    if (err) throw err;
  }).toThrowErrorMatchingSnapshot();
});

test('upOne', async () => {
  await output.upOne(undefined, {silent: true});
  expect(await db.query(sql`SELECT * FROM users`)).toMatchSnapshot();
  await output.upOne(undefined, {silent: true});
  expect(await db.query(sql`SELECT * FROM users`)).toMatchSnapshot();
});
