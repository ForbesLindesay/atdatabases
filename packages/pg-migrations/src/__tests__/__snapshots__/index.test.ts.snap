// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`downAll 1`] = `
"relation \\"users\\" does not exist

> 1 | SELECT * FROM users
    |               ^^^^^
"
`;

exports[`downLast 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Forbes Lindesay",
  },
]
`;

exports[`downLast 2`] = `
Array [
  Object {
    "id": 1,
    "name": "Forbes Lindesay",
  },
]
`;

exports[`downOne 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Forbes Lindesay",
  },
]
`;

exports[`downOne 2`] = `
"relation \\"users\\" does not exist

> 1 | SELECT * FROM users
    |               ^^^^^
"
`;

exports[`generate 1`] = `
"// auto generated by @databases/pg-migrations - do not edit by hand

import packageMigrations, {MigrationsPackage, packageOperation} from '../';

export {MigrationsPackage};
export default packageMigrations(
  {
    id: 'd000jcbsfkyjcwricki6',
    index: 1,
    name: '00001-init.ts',
    operation: async () =>
      packageOperation(await import('./migrations/00001-init')),
  },
  {
    id: 'd000jcbsfkyrhtasvyy8',
    index: 2,
    name: '00002-update.ts',
    operation: async () =>
      packageOperation(await import('./migrations/00002-update')),
  },
);
"
`;

exports[`upAll 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Forbes Lindesay",
  },
  Object {
    "id": 2,
    "name": "Eleanor Brodie",
  },
]
`;

exports[`upOne 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Forbes Lindesay",
  },
]
`;

exports[`upOne 2`] = `
Array [
  Object {
    "id": 1,
    "name": "Forbes Lindesay",
  },
  Object {
    "id": 2,
    "name": "Eleanor Brodie",
  },
]
`;
